apps/emqx/src/emqx_persistent_session_ds_state.erl

```diff
@@ -288,6 +292,39 @@ get_peername(Rec) ->
 set_peername(Val, Rec) ->
     set_meta(?peername, Val, Rec).
 
+-spec get_clientinfo(t()) -> emqx_maybe:t(emqx_types:clientinfo()).
+get_clientinfo(Rec) ->
+    get_meta(?clientinfo, Rec).
+
+-spec set_clientinfo(emqx_types:clientinfo(), t()) -> t().
+set_clientinfo(Val, Rec) ->
+    set_meta(?clientinfo, Val, Rec).
+
+-spec get_will_message(t()) -> emqx_maybe:t(message()).
+get_will_message(Rec) ->
+    get_meta(?will_message, Rec).
+
+-spec set_will_message(emqx_maybe:t(message()), t()) -> t().
+set_will_message(Val, Rec) ->
+    set_meta(?will_message, Val, Rec).
```

Comment:
Keeping the will message inside the `?meta` table/domain will make things harder for the ["session data in DS"](https://github.com/emqx/emqx/pull/12675) feature, because of the lack of transactions there: we'd need to read the metadata and write it back, having the potential of races.

How about we have a table just for that, so that it maps to a topic (in the other PR) that can be safely overwritten without having to read it first?